### PR TITLE
Apply macro to safe spells (part2)

### PR DESCRIPTION
**Pull request by Hosuke**
### Update!
Please build spells in the proper [subproject](../dbt_subprojects/) directory. For more information, please see the main [readme](../README.md), which also links to a GH discussion with the option to ask questions.

### Contribution type
Please check the type of contribution this pull request is for:

- [ ] New spell(s)
- [x] Adding to existing spell lineage
- [ ] Bug fix

**Note:** You can safely discard any section below which doesn't apply based on selection above

---

### For new spell(s)
If you are building new spell(s), please provide the following information:

- **Spell name(s):** *schema.table_name*
- **Description:** [Detailed description of the new spell(s) and their purpose]
- **Who are the new spell(s) for?** [Internal team or general community]
- **How will the new spell(s) be used downstream?** [Description of downstream usage]
- **Implementation details:** [Information on how the spell(s) are implemented]
- **Test instructions:** [How to test the new spell(s)]
- **Related issue(s):** [Link to related issues, if any]

---

### For adding to existing spell lineage
If you are adding to an existing spell lineage, please provide the following information:

- **Description:** [Description of the changes made]

---

### For bug fixes
If you are fixing a bug, please provide the following information:

- **Description:** [Description of the bug fix]
- **Steps to reproduce:** [How to reproduce the bug]
- **Implementation details:** [Information on how the bug was fixed]
- **Test instructions:** [How to test the fix]
- **Related issue(s):** [Link to related issues, if any]

---

### Additional information
Please provide any additional information that might help us review your pull request:

- [Any additional information]

---

Thank you for your contribution!

### CODE REVIEW

**Comment by Hosuke** [comment]
The `bnb` and `polygon` spells cannot be completed within the 90-minute timeframe. All other results are matched, with the exception of the `blast` spell:

https://dune.com/queries/4018595

File: dbt_subprojects/hourly_spellbook/macros/project/safe/safe_transactions.sql [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    tr.output,
+    case
+        when bytearray_substring(tr.input, 1, 4) = 0x6a761202 then 'execTransaction'
+        when bytearray_substring(tr.input, 1, 4) = 0x468721a7 then 'execTransactionFromModule'
+        when bytearray_substring(tr.input, 1, 4) = 0x5229073f then 'execTransactionFromModuleReturnData'
+        else 'unknown'
+    end as method
+from {{ source(blockchain, 'traces') }} tr
+join {{ ref('safe_' ~ blockchain ~ '_safes') }} s
+    on s.address = tr."from"
```
Comment (Hosuke) [inline dbt_subprojects/hourly_spellbook/macros/project/safe/safe_transactions.sql:32]: Retry: join by partition column

File: dbt_subprojects/hourly_spellbook/macros/project/safe/safe_transactions.sql [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+join {{ ref('safe_' ~ blockchain ~ '_singletons') }} ss
+    on tr.to = ss.address
+join {{ source(blockchain, 'transactions') }} et
+    on tr.block_date = et.block_date
+    and tr.tx_hash = et.hash
+    and tr.block_number = et.block_number
```
Comment (jeff-dude) [inline dbt_subprojects/hourly_spellbook/macros/project/safe/safe_transactions.sql:38]:
```suggestion
    and tr.block_number = et.block_number
    {% if is_incremental() %}
    and {{ incremental_predicate('et.block_time') }}
    {% endif %}
```

File: dbt_subprojects/hourly_spellbook/macros/project/safe/safe_transactions.sql [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        0x6a761202, -- execTransaction
+        0x468721a7, -- execTransactionFromModule
+        0x5229073f -- execTransactionFromModuleReturnData
+    )
+    and tr.call_type = 'delegatecall'
+    {% if not is_incremental() %}
+    and tr.block_time > TIMESTAMP '{{ project_start_date }}' -- for initial query optimisation
+    {% else %}
+    and {{ incremental_predicate('tr.block_time') }}
+    and {{ incremental_predicate('et.block_time') }}
```
Comment (jeff-dude) [inline dbt_subprojects/hourly_spellbook/macros/project/safe/safe_transactions.sql:49]:
```suggestion
```

**Review by jeff-dude — CHANGES_REQUESTED** [review]
along with changes below, in order to see run in CI:
- first try updating the param passed into macro for each chain to be `2024-08-01` only for smaller reads
- if that passes CI, and you want to try full history, leverage `prod_exclude` tags to remove some chains that are heavy (bnb, polygon, ethereum) in order to see other chains run. then do the opposite, exclude all other chains but those few, see if they run when cluster is focused on less
  - (this second approach may not be necessary, if we run first small subset of data and ensure it matches prod table where date is filtered to same timeframe)

File: dbt_subprojects/hourly_spellbook/macros/project/safe/safe_transactions.sql [hunk 4]
```diff
@@ -0,0 +1,54 @@
+{% macro safe_transactions(blockchain, project_start_date) %}
+
```
Comment (Hosuke) [inline dbt_subprojects/hourly_spellbook/macros/project/safe/safe_transactions.sql:2]:
```suggestion

-- override param
{% set project_start_date = '2024-07-31' %}

```

File: dbt_subprojects/hourly_spellbook/macros/project/safe/safe_transactions.sql [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+-- override param
+{% set project_start_date = '2024-07-31' %}
```
Comment (Hosuke) [inline dbt_subprojects/hourly_spellbook/macros/project/safe/safe_transactions.sql:4]: Records within the time frame are consistent except for `blast` chain:
https://dune.com/queries/4018595

Comment (jeff-dude) [reply dbt_subprojects/hourly_spellbook/macros/project/safe/safe_transactions.sql:4]: plz do take next steps to dig into blast a bit further on why. is new data more accurate? was there a bug in old code (or present in new code)?

File: dbt_subprojects/hourly_spellbook/models/_project/safe/blast/safe_blast_transactions.sql [hunk 6]
```diff
@@ -17,52 +17,4 @@
...
-        0x468721a7, -- execTransactionFromModule
-        0x5229073f -- execTransactionFromModuleReturnData
-    )
-    and tr.call_type = 'delegatecall'
-    {% if not is_incremental() %}
-    and tr.block_time > TIMESTAMP '{{project_start_date}}' -- for initial query optimisation
-    {% else %}
-    and {{ incremental_predicate('tr.block_time') }}
-    {% endif %}
+{{ safe_transactions('base', project_start_date) }}
```
Comment (Hosuke) [inline dbt_subprojects/hourly_spellbook/models/_project/safe/blast/safe_blast_transactions.sql:20]:
```suggestion
{{ safe_transactions('blast', project_start_date) }}
```

File: dbt_subprojects/hourly_spellbook/macros/project/safe/safe_transactions.sql (hunk 5 above)
Comment (Hosuke) [reply dbt_subprojects/hourly_spellbook/macros/project/safe/safe_transactions.sql:4]: Sorry, typo in the param.

Comment (Hosuke) [reply dbt_subprojects/hourly_spellbook/macros/project/safe/safe_transactions.sql:4]: All matched now:
https://dune.com/queries/4018595

**Review by jeff-dude — APPROVED** [review]
thank you ser 🫡

**Pull request merged**